Automated hardware and benchmarking tests

byteocean opened the pull request:
Automated hardware and benchmarking tests and its detailed usage and configuration can be found in the committed documentation part. Also small fix to the offloading code for issues found during the tests.

byteocean committed:
Automated flow tests using hypervisors and VMs

Rename 'channel' to 'remote machine' and add vm config info wrappers
Imporve code modularity and automate underly IP acquisition

byteocean committed:
Integrate pytest and added basic ipv4 and ipv6 VM2VM tests

byteocean committed:
Added emulated NAT machine test

byteocean committed:
Added LB machine test

byteocean committed:
Do not offload icmp pkts targeting lb address and icmp reply traffic for VM behind NAT

fix #563

byteocean committed:
Added facility to transfer dpservice-cli and generated docker image.

byteocean committed:
Added benchmarking test documentation and optimize the script input parameters

byteocean committed:
Format benchmarking test python code.

byteocean committed:
Ignore benchmarking_test repo when running tap-based tests

PlagueCZ commented on src/nodes/lb_node.c:
```diff
@@ -57,6 +57,9 @@ static __rte_always_inline rte_edge_t get_next_index(__rte_unused struct rte_nod
 				df->nat_type = DP_CHG_UL_DST_IP;
+				cntrack->offload_state.orig = DP_FLOW_OFFLOADED;
+				cntrack->offload_state.reply = DP_FLOW_OFFLOADED;
+				df->offload_state = DP_FLOW_NON_OFFLOAD;
```
If this (and the one-liner in `dnat_node.c`) are a bugfix they should have an issue and the commit message should mark it. Or this should be a separate PR

If it's not a bugfix and something only needed for test (does not seem like), then it should be a conditional compilation.

PlagueCZ commented on docs/testing/automated_benchmark_test.md:
````diff
@@ -0,0 +1,110 @@
+```
+./runtest.py --mode offload --stage cicd --docker-image ghcr.io/ironcore-dev/dpservice:sha-e9b4272 -v
+
+./runtest.py --mode both --stage dev -v
````
Just a thought for maybe easier operations, that we could rename the TAP script to `taptest.py` and this one to `vmtest.py` or something else, just to prevent confusion in the future?

PlagueCZ commented on docs/testing/automated_benchmark_test.md:
```diff
@@ -0,0 +1,110 @@
+
+5. "--env-config-file" and "--env-config-name". They provide information of the above described `test_configurations.json`. It is possible this file is renamed or located somewhere else. And it is also possible to create several configurations within this file and specify one of them for the tests.
+
+6. "--verbose". Specify if pytest runs in the verbose mode to see all steps and results during test execution.
```
Maybe use backticks for the parameters (here and the above sections), because the reuslting markdown with quotes is less readable (can be just me).

PlagueCZ reviewed (commented):
Looks quite big and thorough. I do not have this setup currently and I will be creating one in the near future but need another computer for that.
But since it already has big use-case for testing, that should not stop anything of course.

The only real thing to address for me is the marking of C code change by a simple `fixes XX`.

byteocean committed:
Enhancing benchmark testing suite.

1. rename main entry script to run_benchmarktest.py.
2. use more backticks in docu.
3. allow to grab dpservice log from containers before setup is tore down.
4. optimize tear down sequence to allow ctrl-c at more places
5. use sudo for vm init operations
6. change access config for VM's tmp dir
7. use sudo when adding LB address to VM
8. allow using scp proto to copy files
9. upload file with both sftp and scp protocols

PlagueCZ reviewed (approved):
Looks good, I cannot do proper testing now, but will be in the future so we can update this afterwards

byteocean committed:
Ensure terminating pytest when CTRL-C is hit

Stop VM's ssh.socket before stopping ssh service.

guvenc reviewed (changes requested):
I have successfully tested this script in my environment. Thanks a lot.
I have the following feedback:

1. Please make sure that licence information is added where it is needed to the new files.
2. Ctrl + C during init phase of the script wouldnt stop the script.
3. We need VM provisioning to make the automation, a real automation.
   -  An ideal VM provisioning would then get a pointer to a rsa key pair and vm xml file for each VM and xml file for each default network (virtio) of the hypervisor and together with an ignition file, it can provision the VMs. (with predefined usernames and automatic copy of the ssh keys via ignition)
 
I think second and third point above can be done as a second step. For the details of the third step, we can have a separate meeting to discuss. If the licence part is fixed, this PR can be seen as ready  to be merged, from my point of view.

byteocean committed:
Improve console output in the verbose mode.

byteocean committed:
Improve automated benchmark test script's documentation and add licence headers.

guvenc reviewed (approved):
Looks good to me. Thanks.
